File: 4 kyu/Matrix Determinant/solution.java
public class Matrix {
    
    public static int determinant(int[][] m) {
        int d = 0, size = m.length;
        if (size == 1) return m[0][0];
        
        for (int n = 0 ; n < size ; n++) {
            int[][] newM = new int[size-1][size-1];
            for (int x = 1 ; x < size ; x++) for (int y = 0 ; y < size ; y++) {
                if (y == n) continue;
                newM[x-1][y + (y>n ? -1 : 0)] = m[x][y];
            }
            d += (n%2!=0 ? -1 : 1) * m[0][n] * determinant(newM);
        }
        return d;
    }
}
_____________________________________________
public class Matrix {

    public static int minor(int[][] m, int n, int[] excluded) {
        int size = m.length;
        if (n == size) return 1;
        int sum = 0;
        int sign = 1;
        for (int i = 0; i < size; i++)
            if (excluded[i] == 0) {
                excluded[i] = 1;
                sum += sign * m[n][i] * minor(m, n + 1, excluded);
                sign = - sign;
                excluded[i] = 0;
            }
        return sum;
    }
    
    public static int determinant(int[][] matrix) {
        return minor(matrix, 0, new int[matrix.length]);
    }
}
_____________________________________________
import java.util.Arrays;

public class Matrix {
    
    public static int determinant(int[][] matrix) {
        if (matrix.length == 1) return matrix[0][0];
        
        boolean[] activeCols = new boolean[matrix.length];
        Arrays.fill(activeCols, true);
        return determinant(matrix, 0, activeCols);
    }
    
    private static int determinant(int[][] matrix, int depth,  boolean[] activeCols) {
        if (depth == matrix.length - 1) {
            for (int col = 0; col < matrix.length; col++) {
                if (activeCols[col]) {
                    return matrix[depth][col];
                }
            }
        }
        
        int sign = 1;
        int det = 0;
        for (int col = 0; col < activeCols.length; col++) {
            if (activeCols[col]) {
                activeCols[col] = false;
                det += sign * matrix[depth][col] * determinant(matrix, depth + 1, activeCols);
                activeCols[col] = true;
                
                sign = -sign;
            }
        }
        
        return det;
    }
}
_____________________________________________
public class Matrix{
  public static int determinant(int[][] m){
    if(m.length == 1) return m[0][0];
    int d = 0, k, l;
    for(int j = 0; j < m.length; j++){
      int[][] n = new int[m.length-1][m.length-1];
      for(k = 0; k < m.length; k++) for(l = 0; l < m.length; l++)
        if(k != 0 && l != j) n[k-((k>0)?1:0)][l-((l>j)?1:0)] = m[k][l];
      d += ((j % 2 == 0)?1:-1) * m[0][j] * determinant(n);
    }
    return d;
  }
}

File: 4 kyu/Matrix Determinant/solution.php
function minor(array $matrix, int $x): array
{
    array_shift($matrix);
    foreach ($matrix as $key => $row) {
        $matrix[$key] = array_merge(
            array_slice($row, 0, $x),
            array_slice($row, $x + 1)
        );
    }
    return $matrix;
}

function determinant(array $matrix): int
{
    $size = count($matrix);
    if ($size === 1) {
        return $matrix[0][0];
    }
    
    $result = 0;
    for ($i = 0, $k = 1; $i < $size; $i++, $k = -$k) {
        $result += $k * $matrix[0][$i] * determinant(minor($matrix, $i));
    }
    return $result;
}
_____________________________________________
function determinant(array $matrix): int {
  if (count($matrix) == 1) return $matrix[0][0];
  $d = 0;
  for ($i = 0; $i < count($matrix); $i++) {
    $nm = [];
    for ($j = 1; $j < count($matrix); $j++) {
      $temp = $matrix[$j];
      array_splice($temp, $i, 1);
      array_push($nm, $temp);
    }
    $d += pow(-1, $i) * ($matrix[0][$i]) * determinant($nm);
  }
  return $d;
}
_____________________________________________
//Вычисление det(2x2)
function det(array $matrix): int{ 
  $tmp1=$matrix[0][0]*$matrix[1][1];
  $tmp2=$matrix[0][1]*$matrix[1][0];
  return $tmp1-$tmp2; 
}
//Разложение матрицы до уровня det(2x2)
function decomp_matrix(array $matrix){
  $sum=0;
  $n=count($matrix);    
  for($k=0;$k<$n;$k++){       
          $coef=$matrix[0][$k]; // Коэффициенты
          $mat_tmp=$matrix;            
          for($l=0;$l<$n;$l++){
            unset($mat_tmp[$l][$k]);
            }
          unset($mat_tmp[0]);
          $n_tmp=count($mat_tmp);
          array_splice($mat_tmp,1,0);
          for($t=0;$t<$n_tmp;$t++){
            array_splice($mat_tmp[$t],1,0);
          }          
      // Вычисляем сумму
       if ($n>3){
          $sum_tmp= $coef*decomp_matrix($mat_tmp);  
       }else{
         $sum_tmp=$coef*(det($mat_tmp));    
       }
      //
      if (($k % 2)==0){
        $sum+=$sum_tmp;
      }else {
        $sum-=$sum_tmp;
      }
  } 
  return $sum;
}

function determinant(array $matrix): int {
  $n=count($matrix);
  $sum=0;
  echo "Размер матрицы:".$n."<br>";
  var_export($matrix);
 
 switch ($n){
    case 0:
        $sum=0;
        break;
    case 1:
        $sum=$matrix[0][0];
        break;
    case 2:
        $sum = det($matrix);      
        break;
    default:
        $sum = decomp_matrix($matrix);
        break;
    }  
  return $sum;
}
